designs/SearchCommandSpec.md

```diff
@@ -0,0 +1,96 @@
+
+## Who are the customers
+
+This targets all NuGet.exe and .NET users that want to be able to search for packages from the command line, rather than using NuGet Gallery or VS PMUI. Also, while `nuget.exe list` allows users to search different package sources for packages with a search query, the results are returned sorted by the package Id, rather than by relevance, as proposed for this Search command. Further, the Search command would provide more information on the packages than List currently does.
+
+## Goals
+
+* Providing search functionality for NuGet packages through the commandline for NuGet.exe
+* Implementing a prototype of this feature for dotnet.exe
+* Displaying the search results to users in a way that maximizes the visibility of the most relevant results
+
+## Solution
+
+The search feature leverages the search service provided by the NuGet API. Search targets multiple sources and displays all of their results one after the other. The command line interface is simple: __./NuGet.exe search \<query\>__.
+
+```
+PS C:\> ./NuGet.exe search logging
+```
+
+Each search result displays the name, version, number of downloads, and a preview of the description of the package.
+
+```
+====================
+Source: NuGet.org
+--------------------
+> Microsoft.Extensions.Logging.Abstractions | 5.0.0-preview.6.20305.6 | Downloads: 345,145,935
+Logging abstractions for Microsoft.Extensions.Logging.
+
+Commonly Used Types:
+Microsoft.Extensions.Lo...
+
+--------------------
+> Microsoft.Extensions.Logging | 5.0.0-preview.6.20305.6 | Downloads: 243,186,566
+Logging infrastructure default implementation for Microsoft.Extensions.Logging.
+When using NuGet 3....
+
+--------------------
+> Microsoft.IdentityModel.Logging | 6.6.0 | Downloads: 120,177,247
+Includes Event Source based logging support.
+
+--------------------
+> Microsoft.Extensions.Logging.Configuration | 5.0.0-preview.6.20305.6 | Downloads: 63,205,816
+```
+
+If a particular source returns no results for a query, it will display output that indicates this.
+
+```
+====================
+Source: dotnet-msbuild
+--------------------
+No results found.
+```
+
+### Optional Arguments
+
+Search provides users with some optional arguments that they can use to customize their search queries:
+
+| Name | Description | Usage |
+| ---  |     ---     |  :-:  |
+| PreRelease | `true` or `false` determining whether to include pre-release packages | -PreRelease `<true/false>` |
+| PackageTypes | The package type to use to filter packages. If the provided package type is not a valid package type as defined by the [Package Type document](https://github.com/NuGet/Home/wiki/Package-Type-%5BPacking%5D), an empty result will returned. | -PackageType `<Package Type>`|
```

Comment:
Adoption of package type filtering on the server side is likely very sparse. AFAIK the only implementor is nuget.org. Other feeds like Azure DevOps and GitHub probably don't support it today. Therefore, we should be careful about the wording and graceful failures. There is a way to check if a server supports package type filtering so I would recommend spec'ing here what the behavior is if the server does not support package type filtering. It's up to you, but maybe say "Package type filtering is not supported on this package source." for each source not support it instead of showing any search results at all.

BTW, you detect package type filtering capabilities via the `SearchQueryService/3.5.0` type in the service index, e.g. just search for that string here https://api.nuget.org/v3/index.json and you'll see.

Reply:
```suggestion
| PackageTypes | The package type to use to filter packages (if supported by the source). If the provided package type is not a valid package type as defined by the [Package Type document](https://github.com/NuGet/Home/wiki/Package-Type-%5BPacking%5D), an empty result will returned. | -PackageType `<Package Type>`|
```

Reply:
Should indicate whether the Filter actually got applied.
I think in my ideal world, we wouldn't simply return empty, but instead output:
(at least consider for `-verbosity detailed`)
- "Applying Filters"
- "Ignoring Filters (not supported by source)"
- Error: "Invalid Filter"

Reply:
I've added Donnie's disclaimer in the PackageTypes description, and _-Verbosity detailed_ can provide additional information on the filters being used and whether they were applied or not.